Ensure C global symbols are consistently prefixed

dra27 opened the pull request:
This PR starts out from upstreaming a change which has been carried for some time in [opam-repository-mingw](https://github.com/fdopen/opam-repository-mingw/tree/opam2/packages/ocaml-variants). The compilers there add a prefix to the `list_` functions in `winlist.c`.

- As per https://github.com/ocaml/ocaml/pull/9260/files#r371792438, the Windows runtime functions `win_multi_byte_to_wide_char` and `win_wide_char_to_multi_byte` have the prefix `win_` changed to `caml_`. They're internal - the actual conversions should be done using macros.
- All the remaining functions in the **Unix** implementation of the Unix library are now prefixed `unix_` with the exception of `uerror`.
- The same thing is done to most of the remaining functions in the **Windows** implementation of the Unix library, except that primitives and C functions which were prefixed `win_` have stayed that way (for now).
- `tools/check-symbol-names` has been updated with these changes.

We could change the last remaining functions in win32unix if we wanted. They are: `win_CRT_fd_of_filedescr`, `win_filedescr_of_channel`, `win_handle_fd`, `win_inchannel_of_filedescr`, `win_outchannel_of_filedescr`, `win_clear_close_on_exec`, `win_set_close_on_exec`, `win_create_process`, `win_create_process_native`, `win_terminate_process`,  `win_cleanup`,  `win_startup`,  `win_system`,  `win_alloc_handle`,  `win_alloc_socket`,  `win_set_inherit`,  `win_findclose`,  `win_findfirst`,  `win_findnext`, `win_waitpid`. At present, win32unix specifically is allowed to have `win_` as a prefix.

Thoughts on this so far welcomed. I'm very happy to split off bits which are acceptable (e.g. the two `win_` functions in the runtime) as approved. It won't be a zero-impact change as there are libraries which hack into the internals of the Unix library, but it's a very detectable change - if we're happy to go ahead, we can do some pre-emptive checking in opam-repository for affected packages.

dra27 committed:
Prefix all C symbols in Unix library unix_

dra27 committed:
Handle .refptr. symbols in check-symbol-names

dra27 committed:
Run check-symbol-names on both mingw ports

dra27 committed:
Rename the two Windows unicode C functions

dra27 committed:
Rename val_process_id to caml_win32_process_id

dra27 committed:
Mark unnecessary global symbols from win32unix static

dra27 committed:
Rename win32_maperr to caml_win32_maperr

dra27 committed:
Allow functions in win32unix to begin with win_

kit-ty-kate commented on Changes:
```diff
@@ -209,6 +213,10 @@ OCaml 4.14.0
   WSADuplicateSocket on sockets instead of DuplicateHandle.
   (Antonin Décimo, review by Xavier Leroy and Nicolás Ojeda Bär)
 
+- #10???: Prefix all C functions in the Unix library with `unix_` except for the
```
What about using `caml_unix_` instead?

dra27 replied:
All the other functions already used `unix_`, so it makes it consistent with them. I'm not against that change either, though :)

xavierleroy commented:
Thank you for the clean-up work, it's definitely useful!

You wrote that `uerror` is too widely used  to be renamed.  Is that really so?

In perfect hindsight, for the Unix library, it would make sense to use `caml_unix_` and `caml_win32_` as prefixes instead of `unix_` and `win32_`, just like the Bigarray code now uses `caml_ba_`.

xavierleroy commented:
This PR needs to be rebased and adapted to the recent removal of otherlibs/win32unix.  I'm still mildly in favor of this PR, but the  5.0 window is closing, and we need to focus our changes.

dra27 committed:
Strip GCC emulated TLS mangling

dra27 committed:
Rename win32_socket to caml_win32_socket

dra27 commented:
Rebased - updates to follow.

dra27 committed:
Prefix remaining win32unix-specific functions caml_win32_

dra27 committed:
Prefix C functions in win32unix caml_win32_

Work towards eliminating win_ symbols.

dra27 committed:
Rename win32unix primitives to unix equivalents

Cases where the primitive performed the same function in both unix and
win32unix, but had a different name in the latter.

dra27 committed:
Rename startup and system win32unix primitives

The primitives for Unix library startup and cleanup make sense on Unix,
we just don't need them. The same goes for unix_system (since we use
spawn on Unix). Treat these three primitives like those in Unix which
just aren't used/available on Windows - they're just only present in the
Windows version of the library.

dra27 committed:
Rename win_handle_fd to unix_filedescr_of_fd

dra27 committed:
Rename renaming win_ primitives to unix_

dra27 commented:
I've added some commits which eliminate the use of `win_` as a prefix completely (@MisterDA, @nojb - thoughts + opinions?). In a few cases, we have identically-behaving primitives between what was win32unix and unix but with different names.

Separately, I'll kick off a health-check run to how widespread `uerror` really is and also just how bad renaming all the C symbols to be `caml_unix_` actually would be.

dra27 committed:
Rename uerror to caml_uerror

dra27 committed:
Prefix all unix_ symbols caml_unix_

MisterDA commented:
Seems reasonable to me. I also support the change from `unix_` to `caml_unix`, or more generally to use the `caml_` prefix for all non-static functions.

shindere commented:
David Allsopp (2022/05/25 03:53 -0700):
> I've added some commits which eliminate the use of `win_` as a prefix completely ***@***.***, @nojb - thoughts + opinions?). In a few cases, we have identically-behaving primitives between what was win32unix and unix but with different names.
>
> Separately, I'll kick off a health-check run to how widespread
> `uerror` really is and also just how bad renaming all the C symbols to
> be `caml_unix_` actually would be.

This all sounds really cool! Thanks!

damiendoligez commented:
Even if `uerror` is too widely used to be removed, what about having `caml_uerror` as the official function and `uerror` as a backward-compatible alias that we hope to remove someday?

MisterDA commented:
And perhaps (I don't think we're doing it already) explain in the manual that the `caml_` prefix is reserved for C symbols if that makes sense? I've seen some C stubs in the wild using it.

damiendoligez commented on yacc/wstr.c:
```diff
@@ -22,8 +22,8 @@ static int windows_unicode_enabled = WINDOWS_UNICODE;
 /* Adapted from runtime/win32.c */
-int win_wide_char_to_multi_byte(const wchar_t *s, int slen,
-                                char *out, int outlen)
+int caml_win32_wide_char_to_multi_byte(const wchar_t *s, int slen,
+                                       char *out, int outlen)
```
Do we really want to touch `ocamlyacc`? This is a standalone program that doesn't export any libraries.

dra27 replied:
The only reason was to keep it "in sync" with the function it lifts out of `runtime/win32.c` - I don't have a strong opinion either way 🙂 I would have changed it because grep would have matched the original name.

damiendoligez replied:
Ok, let's keep this change for consistency.

damiendoligez reviewed (approved):
Had a quick look at the changes, nothing amiss except yacc.

My vote also goes to `caml_unix` although this means the PR will touch 150 files instead of 83.

shindere commented:
Damien Doligez (2022/05/25 05:59 -0700):
> @damiendoligez approved this pull request.
>
> Had a quick look at the changes, nothing amiss except yacc.
>
> My vote also goes to `caml_unix` although this means the PR will touch 150 files instead of 83.
>
> > +int caml_win32_wide_char_to_multi_byte(const wchar_t *s, int slen,
> +                                       char *out, int outlen)
>
> Do we really want to touch `ocamlyacc`? This is a standalone program
> that doesn't export any libraries.

It seems to me it's nice to keep it in sync with the rest of the files,
although there is no risk of interferring.

dra27 committed:
Add compatibility macros to socketaddr.h

dra27 commented:
So, just before I pause for a week:
- An opam-health-check run of _this_ PR convinced me that there should be compatibility macros in `socketaddr.h` which I pushed. lwt will require a patch (see https://github.com/dra27/lwt/commit/58437aee8f0825f481dccef05eaa79b0ad0278b3) because it cribs the `socket_domain_table` and `socket_domain_type` values in C. Only one package fails, but more on that in a sec.
- https://github.com/dra27/ocaml/commit/768df9292e192084c1714b7de08bbef248c8419b (not at present in this PR) renames `uerror` to `caml_uerror` (with `unixsupport.h` defining an alias) and https://github.com/dra27/ocaml/commit/fc4f8c3d03dd9bcbafe1395333163f9a0bb2f5e6 then goes the whole hog renaming all the symbols to `caml_unix_`. I tested these separately - only an additional 3 packages break, all for the same reason - it looks like there may be a Mirage library which declares `uerror` itself rather than using `caml/unixsupport.h` and so doesn't pick up the compatibility macro. Lwt needs one extra fix for the full `caml_unix_` rename, because it uses the `unix_exit` primitive (see https://github.com/dra27/lwt/commit/12813c022c99e5c572cfae8ee667ca28855363f7)

However, `ocluster.0.1` fails with all three and it does reveal the limitation of these tests - when we build libraries (both C and OCaml) we don't necessarily try linking programs with them! ocluster reveals internal use of `cst_to_constr` in the extunix package which is otherwise building fine.

I'm not sure this means adding one or both of those additional commits is a bad idea (especially as the fixes where needed in libraries is really quite easy), but a different strategy for bulk testing might be in order. It would be relatively simple to hack Asmlibrarian and Bytelibrarian temporarily to check for the old primitive names, but I won't get to that before 6 June.

Either way, there's still a little bit of time to decide whether this wants to be in alpha1 or not.

dra27 commented:
Rebased - the only change in the commits reviewed already is the addition of a compatibility macro for `code_of_unix_error`:
```diff
diff --git a/otherlibs/unix/unixsupport.h b/otherlibs/unix/unixsupport.h
--- a/otherlibs/unix/unixsupport.h
+++ b/otherlibs/unix/unixsupport.h
@@ -83,6 +83,11 @@ extern void caml_win32_maperr(DWORD errcode);
 extern value unix_error_of_code (int errcode);
 extern int unix_code_of_unix_error (value error);

+/* Compatibility definitions for the pre-5.0 names of these functions */
+#ifndef CAML_BUILDING_UNIX
+#define code_of_unix_error unix_code_of_unix_error
+#endif /* CAML_BUILDING_UNIX */
+
 CAMLnoreturn_start
 extern void unix_error (int errcode, const char * cmdname, value arg)
 CAMLnoreturn_end;
```

and then two extra commits: the first renames `uerror` to `caml_uerror` and then renames _all_ the other symbols to be prefixed `caml_` (so those two commits can be checked separately).

dra27 commented:
For testing, I prepared branches where the type-checker rejects `external` declarations including the old names and I also modified opam-health-check to run `nm` over all the `.a` in an opam switch when testing packages and fail if it spotted C usages of anything.

TL;DR The impact is limited to a very small number of packages, all of which require 5.0 updates already so I think we should do it.

The detail:
- `lwt` uses `socket_domain_table` and `socket_domain_type` so needs a small update to use `caml/version.h` to select the correct symbol. It also uses the `unix_exit` primitive, but it already has preprocessing in place (I likewise have a branch ready)
- `batteries` requires an update, because it uses the `unix_link` primitive (I have a branch ready)
- `core` (up to v0.14) and `core_unix` (from v0.15 - the code moved) require unvendoring `socketaddr.h`, which was done, presumably, for compatibility with OCaml 3.11 and earlier (!!) and also a small tweak using `caml/version.h` to select the correct name for the `unix_{s,g}et_sockopt_aux` (I have branches ready)
- `extunix` and `posixat` use our internal `cst_to_constr` functions so need a small `caml/version.h` tweak (I have a branches ready)
- `mirage-block-unix` and `vhd-format-lwt` (`vhd-format-lwt` contains a vendored copy of `mirage-block-unix`) need updating to use `caml/unixsupport.h` instead of manually declaring `uerror` as an `extern` (I have a branch ready)
- `ocamlnet` requires an update to use `caml/unixsupport.h` instead of declarations (I have a WIP branch ready)
- `win-eventlog` uses `win32_maperr` (curiously, it does that even on a Unix build!)
- `satyrographos` breaks because it's constrained to posixat v0.14.0 (it's already broken on 5.0)
- `qcow-tool breaks` with `uerror` being renamed because it's constrained to mirage-block-unix 2.12.1 (likewise already broken on 5.0)
- `orsetto` and `unison` both use `unix_error` (I haven't done a branch for those yet - they're also broken on 5.0 at the moment)
- `mirage-solo5`, `mirage-clock` and `xapi-stdexit-unix` use Unix primitives directly (similarly haven't done a branch yet - they're also broken on 5.0 already)

The important thing to note is that is that lwt, core/core_unix and extunix required changes to deal with the _basic_ version of this PR - they're not additionally broken by `uerror` and `caml_unix_` prefixing.

damiendoligez commented:
> I think we should do it.

I agree.

damiendoligez commented on otherlibs/unix/socketaddr.h:
```diff
@@ -64,16 +64,25 @@ typedef int socklen_param_type;
+#define get_sockaddr caml_unix_get_sockaddr
+#define alloc_sockaddr caml_unix_alloc_sockaddr
+#define alloc_inet_addr caml_unix_alloc_inet_addr
+#define alloc_inet6_addr caml_unix_alloc_inet6_addr
```
I'd move this one inside the `#ifdef HAS_IPV6` on line 84.

dra27 replied:
Good idea, yes!

```diff
diff --git a/otherlibs/unix/socketaddr.h b/otherlibs/unix/socketaddr.h
index 66e7b02aac..768669d0cc 100644
--- a/otherlibs/unix/socketaddr.h
+++ b/otherlibs/unix/socketaddr.h
@@ -69,7 +69,6 @@ extern "C" {
 #define get_sockaddr caml_unix_get_sockaddr
 #define alloc_sockaddr caml_unix_alloc_sockaddr
 #define alloc_inet_addr caml_unix_alloc_inet_addr
-#define alloc_inet6_addr caml_unix_alloc_inet6_addr
 #endif /* CAML_BUILDING_UNIX */

 extern void caml_unix_get_sockaddr (value mladdr,
@@ -84,7 +83,12 @@ extern value caml_unix_alloc_inet_addr (struct in_addr * inaddr);
 #ifdef HAS_IPV6
 extern value caml_unix_alloc_inet6_addr (struct in6_addr * inaddr);
 #define GET_INET6_ADDR(v) (*((struct in6_addr *) (v)))
-#endif
+
+/* Compatibility definition for the pre-5.0 name of this function */
+#ifndef CAML_BUILDING_UNIX
+#define alloc_inet6_addr caml_unix_alloc_inet6_addr
+#endif /* CAML_BUILDING_UNIX */
+#endif /* HAS_IPV6 */

 #ifdef __cplusplus
 }
```

damiendoligez reviewed (approved):
Looks good except for one misplaced macro.

dra27 commented:
Thanks, @damiendoligez - I also added you to the `Changes` entry on the last push. Good to go with CI, therefore and to 5.0 and alpha0 🥳

damiendoligez commented:
Before merging, wait for @Octachron to finish his own review.

dra27 commented:
ACK

hannesm commented:
This is a great line of work, thanks for working on it.

Octachron reviewed (approved):
I can confirm that the changes look good even with a third pair of eyes.

MisterDA commented on otherlibs/unix/channels_unix.c:
```diff
@@ -34,7 +34,7 @@
    Returns 0 if OK, a nonzero error code if error. */
 
-static int unix_check_stream_semantics(int fd)
+static int caml_unix_check_stream_semantics(int fd)
```
This one doesn't need the `caml_unix_` prefix, it's static, and the Windows file doesn't have it.
(oops, too late).

Octachron replied:
Fortunately, since the function is static, renaming it is not a breaking change.

dra27 replied:
There're a few static symbols which could get simplified names in future, yes! (there's `caml_unix_error_exn` as well). Wood for the trees, 'n all that 🙂

hannesm commented:
Would it be sensible to have a CI job that ensures all global symbols in (e.g.) libasmrun.a contain a `caml_` prefix (by using `readelf` and `grep -v`)? To avoid any accidental inclusion of symbols without such a prefix?

dra27 commented:
Thanks, @hannesm! Thank you also to everyone for the reviewing time on this - these changes in 5.0 are obviously insignificant next to the headline features, but it's very pleasing to have been able to use the major version bump to correct them.

> Would it be sensible to have a CI job that ensures all global symbols in (e.g.) libasmrun.a contain a `caml_` prefix (by using `readelf` and `grep -v`)? To avoid any accidental inclusion of symbols without such a prefix?

There already is 🙂 It was updated as part of this PR. It's done `nm` rather than `readelf` but has the same effect.
